src/main/java/org/opensearch/ad/AnomalyDetectorJobRunner.java

```diff
@@ -468,12 +468,14 @@ private void indexAnomalyResult(
         String detectorId = jobParameter.getName();
         detectorEndRunExceptionCount.remove(detectorId);
         try {
+            // reset error if different from previously recorded one
+            detectionStateHandler.saveError(response.getError(), detectorId);
             // skipping writing to the result index if not necessary
             // For a single-entity detector, the result is not useful if error is null
             // and rcf score (thus anomaly grade/confidence) is null.
-            // For a multi-entity detector, we don't need to save on the detector level.
-            // We always return 0 rcf score if there is no error.
-            if (response.getAnomalyScore() <= 0 && response.getError() == null) {
+            // For a HCAD detector, we don't need to save on the detector level.
+            // We return 0 or Double.NaN rcf score if there is no error.
+            if ((response.getAnomalyScore() <= 0 || Double.isNaN(response.getAnomalyScore())) && response.getError() == null) {
```

Comment:
From the description
> Previously, we won't reset error in the state when a detector changes from an error state to a non-error state. This PR fixes that.

Should we set error as null in this if block?

Reply:
You meant call saveError under this branch?

Reply:
Oh, I see your force pushed a new commit. Not sure what the original code looks like. But current code looks ok for me.